NeoMathEngine/src/CPU/CpuMathEngineDnnRowwise.cpp

```diff
@@ -0,0 +1,250 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#pragma hdrstop
+
+#include <memory>
+#include <vector>
+#include <algorithm>
+
+#include <CpuMathEngine.h>
+#include <CpuMathEngineDnnRowwise.h>
+#include <CpuMathEnginePrivate.h>
+#include <CpuExecutionScope.h>
+#include <MemoryHandleInternal.h>
+
+namespace NeoML {
+
+CRowwiseWrapper::CRowwiseWrapper( float* data, int rowCount, int rowSize ) :
+	firstDataRow( data ),
+	rowCount( rowCount ),
+	rowSize( rowSize ),
+	addedRows( 0 ),
+	removedRows( 0 )
+{}
+
+int CRowwiseWrapper::DataRowCount() const
+{
+	PRESUME_EXPR( addedRows >= removedRows );
+	return addedRows - removedRows;
+}
+
+const float* CRowwiseWrapper::DataRows() const
+{
+	PRESUME_EXPR( DataRowCount() > 0 );
+	return firstDataRow;
+}
+
+float* CRowwiseWrapper::EmptyRows()
+{
+	PRESUME_EXPR( addedRows < rowCount );
+	return firstDataRow + DataRowCount() * rowSize;
+}
+
+void CRowwiseWrapper::AddRows( int count )
+{
+	PRESUME_EXPR( count > 0 );
+	addedRows += count;
+	PRESUME_EXPR( addedRows <= rowCount );
+}
+
+void CRowwiseWrapper::RemoveRows( int count )
+{
+	PRESUME_EXPR( count > 0 );
+	PRESUME_EXPR( count <= DataRowCount() );
+	removedRows += count;
+	firstDataRow += count * rowSize;
+}
+
+CRowwiseBuffer::CRowwiseBuffer( IMathEngine& mathEngine, int rowCount, int rowSize, int fullHeight ) :
+	rowCount( rowCount ),
+	rowSize( rowSize ),
+	fullHeight( fullHeight ),
+	realHeight( std::min( fullHeight, 2 * rowCount ) ),
+	bufferVar( mathEngine, realHeight * rowSize ),
+	bufferPtr( GetRaw( bufferVar.GetHandle() ) ),
+	dataPtr( bufferPtr ),
+	dataPtrIndex( 0 ),
+	dataRowsCount( 0 ),
+	dataRowIndex( 0 )
+{
+}
+
+const float* CRowwiseBuffer::DataRows() const
+{
+	PRESUME_EXPR( dataRowsCount > 0 );
+	return dataPtr;
+}
+
+int CRowwiseBuffer::EmptyRowCount() const
+{
+	return std::min( rowCount - dataRowsCount, realHeight - ( dataPtrIndex + dataRowsCount ) );
+}
+
+float* CRowwiseBuffer::EmptyRows()
+{
+	PRESUME_EXPR( EmptyRowCount() > 0 );
+	return dataPtr + dataRowsCount * rowSize;
+}
+
+void CRowwiseBuffer::AddRows( int count )
+{
+	PRESUME_EXPR( count > 0 );
+	PRESUME_EXPR( count <= EmptyRowCount() );
+	dataRowsCount += count;
+	PRESUME_EXPR( dataRowsCount <= rowCount );
+	PRESUME_EXPR( dataRowsCount + dataPtrIndex <= realHeight );
+}
+
+void CRowwiseBuffer::RemoveRows( int count )
+{
+	PRESUME_EXPR( count > 0 );
+	PRESUME_EXPR( count <= dataRowsCount );
+	dataRowsCount -= count;
+	dataRowIndex += count;
```

Comment:
What's the sense of `dataRowIndex` variable?

Reply:
I've added comments to header.

Reply:
Looks like started index for data rows it's `dataPtrIndex`, not `dataRowIndex`.

Reply:
Hmm, how to explain. There are 2 levels:
1. buffer emulates slice of a bigger data blob. Here is `dataRowIndex`.
2. inside this struct buffer is actually bigger than requested. This done to reduce number of memmoves. Because of this we need the distance between beginning of the allocated memory  and where data starts right now. This is `dataPtrIndex`.